WFS FeatureStore and WFS PagingToolbar

chrismayer opened the pull request:
This PR adds a store to load features from an OGC WFS (`GeoExt.data.store.WfsFeatures`). So WFS features can easily managed and used with common UI components (e.g. grids).

In order to provide a paging mechanism for WFS data a custom WFS PagingToolbar (`GeoExt.toolbar.WfsPaging`) is also introduced by this PR.

See the provided example how to use these two components together.

marcjansen commented:
Looking into this now.

marcjansen commented on examples/features/grid-paging.js:
```diff
@@ -0,0 +1,119 @@
+                        attributions: [new ol.Attribution({
+                            html: '<a href="https://www.openstreetmap.org/' +
+                            'copyright">OpenStreetMap contributors</a>'
+                        })]
```
Can be simplified:

```js
{
  // other properties
  attributions: '<a href="…">…</a>'
  // …
}
```

marcjansen commented on examples/features/grid-paging.js:
```diff
@@ -0,0 +1,119 @@
+            model: 'GeoExt.data.model.Feature',
+            passThroughFilter: true,
+            createLayer: true,
+            url: 'https://maps.dwd.de/geoserver/dwd/ows?',
```
An awesome service! According to the `AccessConstraints` of the [capabilities](https://maps.dwd.de/geoserver/ows?service=wfs&version=2.0.0&request=GetCapabilities) which link to the [Copyright](https://www.dwd.de/DE/service/copyright/copyright_node.html) (German, [English version](https://www.dwd.de/EN/service/copyright/copyright_artikel.html)) we have to attribute correctly (if you follow another link you'll end up [here](https://www.dwd.de/EN/service/copyright/templates_dwd_as_source.html)) like this: `Source: Deutscher Wetterdienst`.

I am unsure whether this is an easy change, but I think we should enable it. Probably in follow-up PR, so this doesn't get to big.

marcjansen commented on examples/features/grid-paging.js:
```diff
@@ -0,0 +1,119 @@
+        });
+        var mapPanel = Ext.create('Ext.panel.Panel', {
+            region: 'center',
+            height: 400,
```
Unneeded:

```suggestion
```

marcjansen commented on src/data/store/WfsFeatures.js:
```diff
@@ -0,0 +1,213 @@
+    typeName: null,
+
+    /**
+     * The 'typeName' param value used in the WFS request.
```
```suggestion
     * The 'outputFormat' param value used in the WFS request.
```

marcjansen commented on src/data/store/WfsFeatures.js:
```diff
@@ -0,0 +1,213 @@
+        if (createLayer) {
+            // the WFS vector layer showing the WFS features on the map
+            me.source = new ol.source.Vector({
+                features: []
```
This is probably the right place for any user defined attributions.

marcjansen commented on src/data/store/WfsFeatures.js:
```diff
@@ -0,0 +1,213 @@
+            method: 'GET',
+            params: params,
+            success: function(response) {
+                var respJson = Ext.decode(response.responseText);
```
Does this assume JSON? I am totally fine with that, but I think we should have it in the API (or maybe I overread it)

weskamm replied:
Please wrap `Ext.decode` in an try catch block, as it will throw when the response is in XML (which happens for invalid request / server problems)

marcjansen commented on src/data/store/WfsFeatures.js:
```diff
@@ -0,0 +1,213 @@
+     * The 'version' param value used in the WFS request.
+     * @cfg {String}
+     */
+    version: '2.0.0',
```
I think that `2.0.0` is also the only real version which makes sense, right? As other versions don't support paging as far as I know. Especially since we assume a certain structure in the success-callback. I am a bit unsure whether it was probably doable with previous versions and some vendor specific params.

marcjansen commented on test/spec/GeoExt/data/store/WfsFeatures.test.js:
```diff
@@ -0,0 +1,166 @@
+    });
+
+    describe('loading without paging', function() {
+        var dataPath = (typeof __karma__ === 'undefined' ? '' : 'base/test/');
```
Under which circumstances is `__karma__` undefined? I never thought this can happen :man_shrugging:

marcjansen commented on test/spec/GeoExt/data/store/WfsFeatures.test.js:
```diff
@@ -0,0 +1,166 @@
+        var store;
+        beforeEach(function() {
+            store = Ext.create('GeoExt.data.store.WfsFeatures', {
+                url: 'foo'
```
Very minor: This adds a warning to the output of  `npm test`. We can also use the URL from the `loading without paging` test, can't we?

marcjansen commented on test/spec/GeoExt/data/store/WfsFeatures.test.js:
```diff
@@ -0,0 +1,166 @@
+        var url = dataPath + 'data/wfs_mock.geojson';
+        beforeEach(function() {
+            div = document.createElement('div');
+            document.body.appendChild(div);
```
Can you make use of `TestUtil` and some methods therein?

chrismayer commented:
Thanks for the reviews guys!

I addressed most of your remarks / suggestions. I also decided to enhance the response parsing, so other formats than JSON are supported. You just have to configure the store with the correct `outputFormat` and the corresponding `ol.format` instance. Feature parsing is now totally done by OL. Regarding the detection of total feature amount for paging I implemented a function which supports GML and GeoJSON. So these two formats should be fully supported now (see 88e14ff) .

marcjansen commented on src/data/store/WfsFeatures.js:
```diff
@@ -0,0 +1,250 @@
+
+    /**
+     * Detects the total amount of features (without paging) of the given
+     * WFS response. The detectioin is based on the response format (currently
```
```suggestion
     * WFS response. The detection is based on the response format (currently
```

marcjansen commented on src/data/store/WfsFeatures.js:
```diff
@@ -0,0 +1,250 @@
+        var me = this;
+        var totalCount = -1;
+        if (me.outputFormat.indexOf('application/json') !== -1) {
+            var respJson = Ext.decode(wfsResponse.responseText);
```
This can still be an XML, right? @weskamm said this would blow up then... maybe `try {} catch (e) {}`?

chrismayer replied:
Correct. Since errors in OGC services are encoded in XML, no matter what you send as outputFormat (:-/), this could blow up. I added a try / catch in 0aa4b81

marcjansen commented on test/spec/GeoExt/data/store/WfsFeatures.test.js:
```diff
@@ -0,0 +1,166 @@
+            }
+            store = null;
+            map = null;
+            TestUtil.teardownTestDiv();
```
```suggestion
            TestUtil.teardownTestDiv(div);
```

chrismayer replied:
Japp, done in 66902

geographika commented:
@chrismayer - this looks excellent. I'll be testing against a MapServer backend today.

chrismayer commented:
Due to feedback of @geographika in https://github.com/meggsimum/cpsi-mapview/issues/95 I added another commit ensuring that the binding of the store's event `gx-wfsstoreload` is done once we receive the final store. The prev. approach had problems when binding the store via ViewModel, because it was bound at a later point. So `gx-wfsstoreload` was never bound.

chrismayer commented:
Hi all, this is still good to go? Then I would merge it. TIA!

chrismayer commented:
Thank you all for your input and reviews :v: I'll merge now.
